src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,374 @@
+
+                return Task.CompletedTask;
+            },
+            this.nodeLifetime.ApplicationStopping,
+            repeatEvery: TimeSpan.FromMinutes(1),
+            startAfter: TimeSpan.FromMinutes(1));
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            ProvenBlockHeader header = null;
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                // Check pending cache first.
+                header = this.GetHeaderFromPendingCache(blockHeight);
+
+                if (header == null)
+                {
+                    // Check store cache next.
+                    header = this.GetHeaderFromStoreCache(blockHeight);
+
+                    if (header == null)
+                    {
+                        // Check the repository (DBreeze).
+                        header = await this.provenBlockHeaderRepository.GetAsync(blockHeight).ConfigureAwait(false);
+
+                        // Add the item to the store cache.
+                        this.Cache.AddOrUpdate(blockHeight, header);
+                    }
+                }
+            }
+
+            if (header != null)
+                this.logger.LogTrace("(-):*.{0}='{1}'", nameof(header), header);
+            else
+                this.logger.LogTrace("(-):null");
+
+            return header;
+        }
+
+        /// <inheritdoc />
+        public async Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            var cachedHeaders = new List<ProvenBlockHeader>();            
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                int heightKey = fromBlockHeight;
+
+                // Check if the items exist in the store cache before checking the repository.
+                ProvenBlockHeader headerCache = null;
+
+                var blockHeightList = new List<int>();
+
+                do
+                {
+                    headerCache = this.GetHeaderFromStoreCache(heightKey);
+
+                    if (headerCache != null)
+                        cachedHeaders.Add(headerCache);
+                    else
+                        blockHeightList.Add(heightKey);
+
+                    heightKey++;
+
+                } while (heightKey <= toBlockHeight);
+
+                // Try and get items from the repository if not found in the store cache.
+                if (blockHeightList.Count > 0)
+                {
+                    var repositoryHeaders = new List<ProvenBlockHeader>();
+
+                    foreach(int blockHeight in blockHeightList)
+                    {
+                        var repositoryHeader = new ProvenBlockHeader();
+
+                        repositoryHeader = await this.provenBlockHeaderRepository.GetAsync(blockHeight).ConfigureAwait(false);
```

Comment:
Was it not the point of calling the range method on the underline repo? currently you just call `provenBlockHeaderRepository.GetAsync(blockHeight)` again instead of `provenBlockHeaderRepository.GetAsync(from, to)`

Reply:
Changed this as @noescape00 mentioned it in his PR comments.